### PR TITLE
Export all light client specific code

### LINKED ISSUE #18: Export logic for constructing light client updates
Currently we are maintaining code for constructing a beefy header in multiple repos, it's better that piece of functionality is consolidated in `beefy-rs`
## Requirements
1. Export the code required to fetch and construct light client updates
2. Update errors with better descriptions.
3. Export all functionality specific to light client that is duplicated in other repos.

### PR DESCRIPTION
**Pull request by Wizdave97**
closes #18

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    let runtime_api = generator.generate_runtime(item_mod, derives);
+    Ok(format!("{}", runtime_api))
+}
+
+#[tokio::main]
+async fn main() -> color_eyre::Result<()> {
+    let build_enabled = env::var("PROVER_ENABLED")
+        .map(|v| v == "1")
+        .unwrap_or(false);
```
Comment (seunlanlege) [inline build.rs:54]: can we use cargo features here?

File: src/queries/helpers.rs [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    };
+
+    Ok(TimeStampExtWithProof { ext, proof })
+}
+
+/// Calculate the proof for the parachain heads added to this leaf
+pub fn calculate_parachain_heads_proof(
+    leaf: Vec<u8>,
+    raw_finalized_para_heads: &BTreeMap<u64, BTreeMap<u32, Vec<u8>>>,
+    para_id: u32,
```
Comment (seunlanlege) [inline src/queries/helpers.rs:87]:
```suggestion
pub fn prove_parachain_headers(
    leaf: Vec<u8>,
    finalized_para_heads: &BTreeMap<u64, BTreeMap<u32, Vec<u8>>>, // what is this structure?
    para_id: u32,
```

File: src/queries/helpers.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        para_head,
+        heads_leaf_index: index.ok_or_else(|| {
+            BeefyClientError::Custom("[get_parachain_headers] heads leaf index is None".to_string())
+        })? as u32,
+        heads_total_count: parachain_leaves.len() as u32,
+    })
+}
+
+/// Get the proof for authority set that signed this commitment
+pub fn get_authority_proof(
```
Comment (seunlanlege) [inline src/queries/helpers.rs:146]:
```suggestion
pub fn prove_authority_set(
```

File: src/queries/relay_chain_queries.rs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+use super::runtime;
+
+pub struct FinalizedParaHeads {
+    pub leaf_indices: Vec<u32>,
+    pub raw_finalized_heads: BTreeMap<u64, BTreeMap<u32, Vec<u8>>>,
+}
+
+/// Get the raw parachain heads finalized in the provided block
+pub async fn get_raw_finalized_parachain_heads<T: Config>(
```
Comment (seunlanlege) [inline src/queries/relay_chain_queries.rs:22]:
```suggestion
pub async fn fetch_finalized_parachain_heads<T: Config>(
```

File: src/queries/relay_chain_queries.rs [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+    }
+
+    Ok(FinalizedParaHeads {
+        raw_finalized_heads: finalized_blocks,
+        leaf_indices,
+    })
+}
+
+/// Get beefy justification for latest finalized beefy block
+pub async fn get_beefy_justifications<T: Config>(
```
Comment (seunlanlege) [inline src/queries/relay_chain_queries.rs:111]:
```suggestion
pub async fn fetch_beefy_justifications<T: Config>(
```

File: src/queries/relay_chain_queries.rs [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        u32,
+        beefy_primitives::crypto::Signature,
+    >::decode(&mut &*beefy_justification)
+    .expect("Beefy justification should decode correctly");
+
+    Ok((signed_commitment, latest_beefy_finalized))
+}
+
+/// Query a batch leaf proof
+pub async fn get_mmr_batch_proof<T: Config>(
```
Comment (seunlanlege) [inline src/queries/relay_chain_queries.rs:150]:
```suggestion
pub async fn fetch_mmr_batch_proof<T: Config>(
```

File: src/queries/relay_chain_queries.rs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        .request(
+            "mmr_generateBatchProof",
+            rpc_params!(leaf_indices, block_hash),
+        )
+        .await?;
+    Ok(proof)
+}
+
+/// Query a single leaf proof
+pub async fn get_mmr_leaf_proof<T: Config>(
```
Comment (seunlanlege) [inline src/queries/relay_chain_queries.rs:167]:
```suggestion
pub async fn fetch_mmr_leaf_proof<T: Config>(
```

**Review by seunlanlege — CHANGES_REQUESTED** [review]
Just needs renaming and some api clarifications

File: build.rs (hunk 1 above)
Comment (Wizdave97) [reply build.rs:54]: Couldn't find anything suggesting we could.

**Review by seunlanlege — APPROVED** [review]

**Pull request merged**